### ISSUE TITLE
train_model works with mallet from CRAN, but not mallet from github

### ISSUE THREAD

**Issue by cbdavis** [agoldst/dfrtopics#3]
The latest version of the [R mallet package on github](https://github.com/mimno/RMallet) causes some problems with the `train_model` function of `R/model.R`.  The main issue seems to be that `trainer$model` isn't defined in the latest github version of the mallet code, but it is on the CRAN version.

To duplicate:

Remove existing mallet package, and re-install from github:

```
remove.packages("mallet")
library(devtools)
install_github("mimno/RMallet/mallet")
```

Create some instances and then try to train a model:

```
m <- train_model(instance, n_topics=50, n_iters=300, metadata=df, threads=4)
```

...and then the error:

```
Apr 26, 2016 11:41:36 AM cc.mallet.topics.ParallelTopicModel <init>
INFO: Mallet LDA: 50 topics, 6 topic bits, 111111 topic mask
Error in trainer$model : no field, method or inner class called 'model' 
```

I'm able to fix this by just using the version from CRAN:

```
remove.packages("mallet")
install.packages("mallet")
```

So this works for now, until the CRAN version gets updated to the latest github version.

**Comment by agoldst** [comment]
Thanks for the careful documentation. I've known about this issue for a while. In the latest MALLET (included in the github R mallet) RTopicModel is a subclass of ParallelTopicModel, whereas CRAN mallet has an older MALLET version in which RTopicModel has a ParallelTopicModel as a member. I have an update to dfrtopics for this but I've been waiting for R mallet to be updated on CRAN, so that installing my own package would be simpler. However, since I've been waiting for a while--the mallet maintainer has a lot on his plate!--I think it would be better to just have my package check for both possibilities. The github R mallet has other features I like to use too. Will work on this.

**Comment by agoldst** [comment]
See how you fare with the latest version on master here. With this update, on my system I can use `train_model` etc. with either the github `mallet` or the CRAN `mallet`. I'll play with it a little more before declaring this fixed, so let me know if you still have issues (or develop new ones).

**Comment by cbdavis** [comment]
It looks good although the one issue I'm trying to track down relates to having multiple threads.  If I run the latest `dfrtopics` + the latest `mallet` from github, then it seems that only a single thread is running, whereas with the CRAN version of `mallet` and the previous version of `dfrtopics`, I could max out my CPUs by specifying `threads=8` in `train_model`.

**Comment by agoldst** [comment]
Thanks. I'm not sure whether I can reproduce this. Is it a `mallet` issue or a `dfrtopics` issue? If you keep the new `dfrtopics` but revert to CRAN `mallet` do you get the behavior you expect? Or if you build command-line `mallet` from the [github version](mimno/Mallet) and do `mallet train-topics --num-threads 8`? As far as I can tell `dfrtopics::train_model` with the most recent github versions successfully sets the model's `numThreads` member to whatever you tell it to be and the number of OS threads for my `rsession` process goes up accordingly. If I rebuild mallet with explicit logging about the threads it claims to be "submitting" them all at each iteration, whatever that means.

**Comment by cbdavis** [comment]
I think it was an issue on my end related to the amount of text used in testing.  I ran some tests with different combinations of versions of `dfrtopics` & `mallet` and I'm able to get multiple threads running, provided that I use enough text as input.  In the code below, if you add or remove `text = rep(text, 10)` you can see the difference in operation.

```
# Test using Pride and Prejudice
if (!file.exists("pg1342.txt")){
  download.file("http://www.gutenberg.org/cache/epub/1342/pg1342.txt", "pg1342.txt")  
}
text = readLines("pg1342.txt") # every 'document' for mallet is a line of text
text = text[which(!text == "")] # remove empty lines
# it seems that multiple threads are not used unless we have a certain size of text, 
# so make some more text
text = rep(text, 10)

library(mallet)
instance = mallet.import(as.character(c(1:length(text))), text, "en.txt")
m <- train_model(instance, n_topics=50, n_iters=100, threads=8)
```

**Comment by agoldst** [comment]
Good--I'm glad things seem to be working. I'm pretty sure mallet won't start more threads than documents (sensibly enough). If you use the package to do anything you're willing to share, I'd love to know: e-mail is first name dot last name at rutgers dot edu.

**Comment by cbdavis** [comment]
I think the multiple threads may be based on the amount of text to be analyzed, rather than the number of documents.  I'm indeed using a single book for testing, but I first split it up based on lines of text, which gives me > 10,000 'documents' to import into mallet, even without `text = rep(text, 10)`.  In my initial testing, I think I got confused since I did a test involving 1000 journal abstracts (300-500 words each) and this wasn't enough to make `mallet` use multiple threads.
